src/DacpacTool/ScriptParser.cs

```diff
@@ -0,0 +1,86 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Linq;
+using System.Reflection;
+using Microsoft.SqlTools.ServiceLayer.BatchParser;
+using Microsoft.SqlTools.Extensibility;
+
+namespace MSBuild.Sdk.SqlProj.DacpacTool
+{
+    public class ScriptParser : ICommandHandler
+    {
+        private readonly Parser _parser;
+        private bool _parsed;
+        private readonly List<string> _includedFileNames = new List<string>();
+        private StringBuilder _scriptBuilder = new StringBuilder();
+
+        public ScriptParser(string sourceFile, IVariableResolver variableResolver)
+        {
+            _parser = new Parser(this, variableResolver, new StreamReader(sourceFile), sourceFile);
+        }
+
+        private void Parse()
+        {
+            if (! _parsed)
+            {
+                _parser.Parse();
+                _parsed = true;
+            }
+        }
+
+        public List<string> CollectFileNames()
+        {
+            Parse();
+            return _includedFileNames;
+        }
+
+        public string GenerateScript()
+        {
+            Parse();
+            return _scriptBuilder.ToString();
+        }
+
+        BatchParserAction ICommandHandler.Go(TextBlock batch, int repeatCount, SqlCmdCommand tokenType)
+        {
+            batch.GetText(true, out string batchText, out LineInfo lineInfo);
+            _scriptBuilder.Append(batchText);
+            _scriptBuilder.Append("\nGO\n");
+            return BatchParserAction.Continue;
+        }
+
+        BatchParserAction ICommandHandler.Include(TextBlock includeBlock, out TextReader stream, out string newFilename)
+        {
+            includeBlock.GetText(true, out string includedFileName, out LineInfo lineInfo);
+
+            if (!Path.IsPathRooted(includedFileName))
+            {
+                var position = lineInfo.GetStreamPositionForOffset(0);
+                includedFileName = Path.Combine(Path.GetDirectoryName(position.Filename), includedFileName);
+            }
+
+            _includedFileNames.Add(includedFileName);
+
+            stream = new StreamReader(includedFileName);
+            newFilename = includedFileName;
+            return BatchParserAction.Continue;
+        }
+
+        BatchParserAction ICommandHandler.OnError(Token token, OnErrorAction action)
+        {
+            // Write error to console
+            Console.Error.WriteLine($"Error encountered in {token.Filename}, line {token.Begin.Line}");
```

Comment:
Could we format the output similar to how the recently added `ModelValidationError` class does it? That would at least give users traceability to where the error is occurring. We might have to come up with our own error code though. I'm wondering how SSDT handles these errors. Could you try to trigger an error with that and see how that is handled?

Reply:
@jmezach @ErikEJ I've pushed a change for this, but I'm not sure this is how we want to handle it... for now, I've basically copied the `StringBuilder` code from `ModelValidationError` into this method. I'd rather reuse `ModelValidationError`, but I don't think I can instantiate a `DacModelError`, since it doesn't have a documented constructor.

Another option would be to convert `ModelValidationError` to have a bunch of private fields that get copied in from the `DacModelError` on instantiation, then provide an overload that allows all of those fields to be passed as individual arguments, but thought I'd run it by you guys before making a change like that.

Reply:
I think that makes sense. When I added the `ModelValidationError` class I just designed it to fix the problem at hand, but if we're going to use it in other places I don't mind needing to do some copying of that data from the `DacModelError`.

Reply:
Great, I've pushed that change

Reply:
This got added too